--- pgtap.sql	2010-05-21 17:56:54.000000000 -0400
+++ pgtap.sql.new	2010-05-21 17:56:36.000000000 -0400
@@ -5588,7 +5588,7 @@
 CREATE OR REPLACE FUNCTION _runem( text[], boolean )
 RETURNS SETOF TEXT AS $$
 DECLARE
-    tap    text;
+    rec    record;
     lbound int := array_lower($1, 1);
 BEGIN
     IF lbound IS NULL THEN RETURN; END IF;
@@ -5596,8 +5596,8 @@
         -- Send the name of the function to diag if warranted.
         IF $2 THEN RETURN NEXT diag( $1[i] || '()' ); END IF;
         -- Execute the tap function and return its results.
-        FOR tap IN EXECUTE 'SELECT * FROM ' || $1[i] || '()' LOOP
-            RETURN NEXT tap;
+        FOR rec IN EXECUTE 'SELECT * FROM ' || $1[i] || '() AS b(a)' LOOP
+            RETURN NEXT rec.a;
         END LOOP;
     END LOOP;
     RETURN;
@@ -5659,14 +5659,14 @@
     setup    ALIAS FOR $3;
     teardown ALIAS FOR $4;
     tests    ALIAS FOR $5;
-    tap      text;
+    rec      record;
     verbos   boolean := _is_verbose(); -- verbose is a reserved word in 8.5.
     num_faild INTEGER := 0;
 BEGIN
     BEGIN
         -- No plan support.
         PERFORM * FROM no_plan();
-        FOR tap IN SELECT * FROM _runem(startup, false) LOOP RETURN NEXT tap; END LOOP;
+        FOR rec IN SELECT * FROM _runem(startup, false) AS b(a) LOOP RETURN NEXT rec.a; END LOOP;
     EXCEPTION
         -- Catch all exceptions and simply rethrow custom exceptions. This
         -- will roll back everything in the above block.
@@ -5681,15 +5681,15 @@
                 IF verbos THEN RETURN NEXT diag(tests[i] || '()'); END IF;
 
                 -- Run the setup functions.
-                FOR tap IN SELECT * FROM _runem(setup, false) LOOP RETURN NEXT tap; END LOOP;
+                FOR rec IN SELECT * FROM _runem(setup, false) AS b(a) LOOP RETURN NEXT rec.a; END LOOP;
 
                 -- Run the actual test function.
-                FOR tap IN EXECUTE 'SELECT * FROM ' || tests[i] || '()' LOOP
-                    RETURN NEXT tap;
+                FOR rec IN EXECUTE 'SELECT * FROM ' || tests[i] || '() AS b(a)' LOOP
+                    RETURN NEXT rec.a;
                 END LOOP;
 
                 -- Run the teardown functions.
-                FOR tap IN SELECT * FROM _runem(teardown, false) LOOP RETURN NEXT tap; END LOOP;
+                FOR rec IN SELECT * FROM _runem(teardown, false) AS b(a) LOOP RETURN NEXT rec.a; END LOOP;
 
                 -- Remember how many failed and then roll back.
                 num_faild := num_faild + num_failed();
@@ -5704,7 +5704,7 @@
         END LOOP;
 
         -- Run the shutdown functions.
-        FOR tap IN SELECT * FROM _runem(shutdown, false) LOOP RETURN NEXT tap; END LOOP;
+        FOR rec IN SELECT * FROM _runem(shutdown, false) AS b(a) LOOP RETURN NEXT rec.a; END LOOP;
 
         -- Raise an exception to rollback any changes.
         RAISE EXCEPTION '__TAP_ROLLBACK__';
@@ -5715,8 +5715,8 @@
         END IF;
     END;
     -- Finish up.
-    FOR tap IN SELECT * FROM _finish( currval('__tresults___numb_seq')::integer, 0, num_faild ) LOOP
-        RETURN NEXT tap;
+    FOR rec IN SELECT * FROM _finish( currval('__tresults___numb_seq')::integer, 0, num_faild ) AS b(a) LOOP
+        RETURN NEXT rec.a;
     END LOOP;
 
     -- Clean up and return.
@@ -6947,7 +6947,7 @@
 DECLARE
     rec    RECORD;
 BEGIN
-    EXECUTE _query($1) INTO rec;
+    FOR rec in EXECUTE _query($1) LOOP END LOOP;
     IF NOT textin(record_out(rec)) IS DISTINCT FROM textin(record_out($2))
         THEN RETURN ok(true, $3);
     END IF;
