--- sql/pgtap.sql
+++ sql/pgtap.sql
@@ -2483,7 +2483,7 @@
              || p.prorettype::regtype AS returns,
            p.prolang     AS langoid,
            p.proisstrict AS is_strict,
-           p.prokind     AS kind,
+           p.proisagg    AS is_agg,
            p.prosecdef   AS is_definer,
            p.proretset   AS returns_set,
            p.provolatile::char AS volatility,
@@ -5656,7 +5656,7 @@
 
 CREATE OR REPLACE FUNCTION _agg ( NAME, NAME, NAME[] )
 RETURNS BOOLEAN AS $$
-    SELECT kind = 'a'
+    SELECT is_agg
       FROM tap_funky
      WHERE schema = $1
        AND name   = $2
@@ -5665,12 +5665,12 @@
 
 CREATE OR REPLACE FUNCTION _agg ( NAME, NAME )
 RETURNS BOOLEAN AS $$
-    SELECT kind = 'a' FROM tap_funky WHERE schema = $1 AND name = $2
+    SELECT is_agg FROM tap_funky WHERE schema = $1 AND name = $2
 $$ LANGUAGE SQL;
 
 CREATE OR REPLACE FUNCTION _agg ( NAME, NAME[] )
 RETURNS BOOLEAN AS $$
-    SELECT kind = 'a'
+    SELECT is_agg
       FROM tap_funky
      WHERE name = $1
        AND args = array_to_string($2, ',')
@@ -5679,7 +5679,7 @@
 
 CREATE OR REPLACE FUNCTION _agg ( NAME )
 RETURNS BOOLEAN AS $$
-    SELECT kind = 'a' FROM tap_funky WHERE name = $1 AND is_visible;
+    SELECT is_agg FROM tap_funky WHERE name = $1 AND is_visible;
 $$ LANGUAGE SQL;
 
 -- is_aggregate( schema, function, args[], description )
