--- pgtap.sql.orig	2009-07-01 10:24:04.000000000 -0700
+++ pgtap.sql	2009-07-01 10:35:45.000000000 -0700
@@ -5780,8 +5780,9 @@
         SELECT pg_catalog.format_type(a.atttypid, a.atttypmod)
           FROM pg_catalog.pg_attribute a
           JOIN pg_catalog.pg_class c ON a.attrelid = c.oid
+          JOIN pg_catalog.pg_namespace n ON c.relnamespace = n.oid
          WHERE c.relname = $1
-           AND c.relistemp
+           AND n.nspname LIKE 'pg_temp%'
            AND attnum > 0
            AND CASE WHEN attisdropped THEN false ELSE pg_type_is_visible(a.atttypid) END
          ORDER BY attnum
@@ -5885,7 +5886,7 @@
     FETCH have INTO rec_have;
     FETCH want INTO rec_want;
     WHILE rec_have IS NOT NULL OR rec_want IS NOT NULL LOOP
-        IF rec_have IS DISTINCT FROM rec_want THEN
+        IF rec_have::text IS DISTINCT FROM rec_want::text THEN
             RETURN ok( false, $3 ) || E'\n' || diag(
                 '    Results differ beginning at row ' || rownum || E':\n' ||
                 '        have: ' || CASE WHEN rec_have IS NULL THEN '()' ELSE rec_have::text END || E'\n' ||
