--- sql/pgtap.sql
+++ sql/pgtap.sql
@@ -784,12 +784,7 @@
     GET STACKED DIAGNOSTICS
         detail  = PG_EXCEPTION_DETAIL,
         hint    = PG_EXCEPTION_HINT,
-        context = PG_EXCEPTION_CONTEXT,
-        schname = SCHEMA_NAME,
-        tabname = TABLE_NAME,
-        colname = COLUMN_NAME,
-        chkname = CONSTRAINT_NAME,
-        typname = PG_DATATYPE_NAME;
+        context = PG_EXCEPTION_CONTEXT;
     RETURN ok( FALSE, descr ) || E'\n' || diag(
            '    died: ' || _error_diag(SQLSTATE, SQLERRM, detail, hint, context, schname, tabname, colname, chkname, typname)
     );
@@ -6696,12 +6691,7 @@
                 GET STACKED DIAGNOSTICS
                     detail  = PG_EXCEPTION_DETAIL,
                     hint    = PG_EXCEPTION_HINT,
-                    context = PG_EXCEPTION_CONTEXT,
-                    schname = SCHEMA_NAME,
-                    tabname = TABLE_NAME,
-                    colname = COLUMN_NAME,
-                    chkname = CONSTRAINT_NAME,
-                    typname = PG_DATATYPE_NAME;
+                    context = PG_EXCEPTION_CONTEXT;
             END;
 
             -- Always raise an exception to rollback any changes.
