--- sql/pgtap.sql
+++ sql/pgtap.sql
@@ -3585,7 +3585,7 @@ RETURNS TEXT AS $$
                AND n.nspname = $1
                AND t.typname = $2
                AND t.typtype = 'e'
-             ORDER BY e.enumsortorder
+             ORDER BY e.oid
         ),
         $3,
         $4
@@ -3613,7 +3613,7 @@ RETURNS TEXT AS $$
                AND pg_catalog.pg_type_is_visible(t.oid)
                AND t.typname = $1
                AND t.typtype = 'e'
-             ORDER BY e.enumsortorder
+             ORDER BY e.oid
         ),
         $2,
         $3
