--- sql/pgtap.sql
+++ sql/pgtap.sql
@@ -7418,7 +7418,6 @@
               JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
              WHERE n.nspname = $1
                AND c.relname = $2
-               AND NOT t.tgisinternal
             EXCEPT
             SELECT $3[i]
               FROM generate_series(1, array_upper($3, 1)) s(i)
@@ -7433,7 +7432,6 @@
               JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
              WHERE n.nspname = $1
                AND c.relname = $2
-               AND NOT t.tgisinternal
         ),
         $4
     );
@@ -7457,7 +7455,6 @@
               JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
              WHERE c.relname = $1
                AND n.nspname NOT IN ('pg_catalog', 'information_schema')
-               AND NOT t.tgisinternal
             EXCEPT
             SELECT $2[i]
               FROM generate_series(1, array_upper($2, 1)) s(i)
@@ -7471,7 +7468,6 @@
               JOIN pg_catalog.pg_class c ON c.oid = t.tgrelid
               JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
                AND n.nspname NOT IN ('pg_catalog', 'information_schema')
-               AND NOT t.tgisinternal
         ),
         $3
     );
