--- pgtap.sql.orig	2009-12-16 16:16:23.000000000 -0800
+++ pgtap.sql	2009-12-16 16:17:15.000000000 -0800
@@ -15,6 +15,11 @@
 RETURNS text AS 'SELECT current_setting(''server_version'')'
 LANGUAGE SQL IMMUTABLE;
 
+CREATE OR REPLACE FUNCTION pg_typeof("any")
+RETURNS regtype
+AS '$libdir/pgtap'
+LANGUAGE C STABLE;
+
 CREATE OR REPLACE FUNCTION pg_version_num()
 RETURNS integer AS $$
     SELECT s.a[1]::int * 10000
@@ -5849,8 +5854,9 @@
         SELECT display_type(a.atttypid, a.atttypmod)
           FROM pg_catalog.pg_attribute a
           JOIN pg_catalog.pg_class c ON a.attrelid = c.oid
+          JOIN pg_catalog.pg_namespace n ON c.relnamespace = n.oid
          WHERE c.relname = $1
-           AND c.relistemp
+           AND n.nspname LIKE 'pg_temp%'
            AND attnum > 0
            AND CASE WHEN attisdropped THEN false ELSE pg_type_is_visible(a.atttypid) END
          ORDER BY attnum
@@ -6197,7 +6203,7 @@
     FETCH want INTO want_rec;
     want_found := FOUND;
     WHILE have_found OR want_found LOOP
-        IF have_rec IS DISTINCT FROM want_rec OR have_found <> want_found THEN
+        IF have_rec::text IS DISTINCT FROM want_rec::text OR have_found <> want_found THEN
             RETURN ok( false, $3 ) || E'\n' || diag(
                 '    Results differ beginning at row ' || rownum || E':\n' ||
                 '        have: ' || CASE WHEN have_found THEN have_rec::text ELSE 'NULL' END || E'\n' ||
@@ -6352,7 +6358,7 @@
     FETCH want INTO want_rec;
     want_found := FOUND;
     WHILE have_found OR want_found LOOP
-        IF have_rec IS DISTINCT FROM want_rec OR have_found <> want_found THEN
+        IF have_rec::text IS DISTINCT FROM want_rec::text OR have_found <> want_found THEN
             RETURN ok( true, $3 );
         ELSE
             FETCH have INTO have_rec;
@@ -6538,13 +6544,7 @@
 $$ LANGUAGE sql;
 
 -- collect_tap( tap, tap, tap )
-CREATE OR REPLACE FUNCTION collect_tap( VARIADIC text[] )
-RETURNS TEXT AS $$
-    SELECT array_to_string($1, E'\n');
-$$ LANGUAGE sql;
-
--- collect_tap( tap[] )
-CREATE OR REPLACE FUNCTION collect_tap( VARCHAR[] )
+CREATE OR REPLACE FUNCTION collect_tap( text[] )
 RETURNS TEXT AS $$
     SELECT array_to_string($1, E'\n');
 $$ LANGUAGE sql;
