--- sql/pgtap.sql
+++ sql/pgtap.sql
@@ -7473,7 +7473,6 @@
               JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
              WHERE n.nspname = $1
                AND c.relname = $2
-               AND NOT t.tgisinternal
             EXCEPT
             SELECT $3[i]
               FROM generate_series(1, array_upper($3, 1)) s(i)
@@ -7488,7 +7487,6 @@
               JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
              WHERE n.nspname = $1
                AND c.relname = $2
-               AND NOT t.tgisinternal
         ),
         $4
     );
@@ -7512,7 +7510,6 @@
               JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
              WHERE c.relname = $1
                AND n.nspname NOT IN ('pg_catalog', 'information_schema')
-               AND NOT t.tgisinternal
             EXCEPT
             SELECT $2[i]
               FROM generate_series(1, array_upper($2, 1)) s(i)
@@ -7526,7 +7523,6 @@
               JOIN pg_catalog.pg_class c ON c.oid = t.tgrelid
               JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
                AND n.nspname NOT IN ('pg_catalog', 'information_schema')
-               AND NOT t.tgisinternal
         ),
         $3
     );
