--- uninstall_pgtap.sql.orig	2011-02-01 14:58:51.000000000 -0800
+++ uninstall_pgtap.sql	2011-02-01 14:58:51.000000000 -0800
@@ -58,8 +58,7 @@
 DROP FUNCTION throws_like ( TEXT, TEXT );
 DROP FUNCTION throws_like ( TEXT, TEXT, TEXT );
 DROP FUNCTION _tlike ( BOOLEAN, TEXT, TEXT, TEXT );
-DROP FUNCTION collect_tap( VARCHAR[] );
-DROP FUNCTION collect_tap( VARIADIC text[] );
+DROP FUNCTION collect_tap( text[] );
 DROP FUNCTION is_empty( TEXT );
 DROP FUNCTION is_empty( TEXT, TEXT );
 DROP FUNCTION isa_ok( anyelement, regtype );
@@ -701,6 +700,7 @@
 DROP FUNCTION _get ( text );
 DROP FUNCTION no_plan();
 DROP FUNCTION plan( integer );
+DROP FUNCTION pg_typeof("any");
 DROP FUNCTION pgtap_version();
 DROP FUNCTION os_name();
 DROP FUNCTION pg_version_num();
