--- uninstall_pgtap.sql.in	2009-05-28 11:16:45.000000000 -0700
+++ uninstall_pgtap.sql	2009-05-28 11:24:32.000000000 -0700
@@ -243,10 +243,6 @@
 DROP FUNCTION has_role( NAME );
 DROP FUNCTION has_role( NAME, TEXT );
 DROP FUNCTION _has_role( NAME );
-DROP FUNCTION enum_has_labels( NAME, NAME[] );
-DROP FUNCTION enum_has_labels( NAME, NAME[], TEXT );
-DROP FUNCTION enum_has_labels( NAME, NAME, NAME[] );
-DROP FUNCTION enum_has_labels( NAME, NAME, NAME[], TEXT );
 DROP FUNCTION hasnt_enum( NAME );
 DROP FUNCTION hasnt_enum( NAME, TEXT );
 DROP FUNCTION hasnt_enum( NAME, NAME );
