--- sql/uninstall_pgtap.sql
+++ sql/uninstall_pgtap.sql
@@ -169,8 +169,7 @@
 DROP FUNCTION throws_like ( TEXT, TEXT );
 DROP FUNCTION throws_like ( TEXT, TEXT, TEXT );
 DROP FUNCTION _tlike ( BOOLEAN, TEXT, TEXT, TEXT );
-DROP FUNCTION collect_tap( VARCHAR[] );
-DROP FUNCTION collect_tap( VARIADIC text[] );
+DROP FUNCTION collect_tap( text[] );
 DROP FUNCTION isnt_empty( TEXT );
 DROP FUNCTION isnt_empty( TEXT, TEXT );
 DROP FUNCTION is_empty( TEXT );
@@ -845,6 +844,7 @@
 DROP FUNCTION _get ( text );
 DROP FUNCTION no_plan();
 DROP FUNCTION plan( integer );
+DROP FUNCTION pg_typeof("any");
 DROP FUNCTION pgtap_version();
 DROP FUNCTION os_name();
 DROP FUNCTION pg_version_num();
