--- uninstall_pgtap.sql.orig	2009-12-16 16:16:49.000000000 -0800
+++ uninstall_pgtap.sql	2009-12-16 16:20:53.000000000 -0800
@@ -57,8 +57,7 @@
 DROP FUNCTION throws_like ( TEXT, TEXT );
 DROP FUNCTION throws_like ( TEXT, TEXT, TEXT );
 DROP FUNCTION _tlike ( BOOLEAN, TEXT, TEXT, TEXT );
-DROP FUNCTION collect_tap( VARCHAR[] );
-DROP FUNCTION collect_tap( VARIADIC text[] );
+DROP FUNCTION collect_tap( text[] );
 DROP FUNCTION is_empty( TEXT );
 DROP FUNCTION is_empty( TEXT, TEXT );
 DROP FUNCTION isa_ok( anyelement, regtype );
@@ -698,6 +697,7 @@
 DROP FUNCTION _get ( text );
 DROP FUNCTION no_plan();
 DROP FUNCTION plan( integer );
+DROP FUNCTION pg_typeof("any");
 DROP FUNCTION pgtap_version();
 DROP FUNCTION os_name();
 DROP FUNCTION pg_version_num();
