--- uninstall_pgtap.sql.orig	2009-12-18 09:53:39.000000000 -0800
+++ uninstall_pgtap.sql	2009-12-18 09:55:43.000000000 -0800
@@ -359,11 +359,6 @@
 DROP FUNCTION has_user( NAME );
 DROP FUNCTION has_user( NAME, TEXT );
 DROP FUNCTION _is_super( NAME );
-DROP FUNCTION hasnt_role( NAME );
-DROP FUNCTION hasnt_role( NAME, TEXT );
-DROP FUNCTION has_role( NAME );
-DROP FUNCTION has_role( NAME, TEXT );
-DROP FUNCTION _has_role( NAME );
 DROP FUNCTION hasnt_enum( NAME );
 DROP FUNCTION hasnt_enum( NAME, TEXT );
 DROP FUNCTION hasnt_enum( NAME, NAME );
@@ -710,5 +705,9 @@
 DROP FUNCTION textarray_text(text[]);
 DROP CAST (boolean AS char(1));
 DROP FUNCTION booltext(boolean);
+DROP CAST (int2vector AS int[]);
+DROP FUNCTION int2vint(int2vector);
+DROP CAST (oidvector AS regtype[]);
+DROP FUNCTION oidvregtype(oidvector);
 -- ## SET search_path TO public;
 -- ## DROP SCHEMA TAPSCHEMA;
