--- uninstall_pgtap.sql.orig	2009-05-28 15:35:12.000000000 -0700
+++ uninstall_pgtap.sql	2009-05-28 15:35:12.000000000 -0700
@@ -250,11 +250,6 @@
 DROP FUNCTION has_user( NAME );
 DROP FUNCTION has_user( NAME, TEXT );
 DROP FUNCTION _is_super( NAME );
-DROP FUNCTION hasnt_role( NAME );
-DROP FUNCTION hasnt_role( NAME, TEXT );
-DROP FUNCTION has_role( NAME );
-DROP FUNCTION has_role( NAME, TEXT );
-DROP FUNCTION _has_role( NAME );
 DROP FUNCTION hasnt_enum( NAME );
 DROP FUNCTION hasnt_enum( NAME, TEXT );
 DROP FUNCTION hasnt_enum( NAME, NAME );
