--- sql/uninstall_pgtap.sql
+++ sql/uninstall_pgtap.sql
@@ -1,9 +1,3 @@
-DROP FUNCTION extensions_are( NAME[] )
-DROP FUNCTION extensions_are( NAME[], TEXT );
-DROP FUNCTION extensions_are( NAME, NAME[] );
-DROP FUNCTION extensions_are( NAME, NAME[], TEXT );
-DROP FUNCTION _extensions();
-DROP FUNCTION _extensions( NAME );
 DROP FUNCTION foreign_tables_are ( NAME[] );
 DROP FUNCTION foreign_tables_are ( NAME, NAME[] );
 DROP FUNCTION foreign_tables_are ( NAME[], TEXT );
