--- uninstall_pgtap.sql.orig	2009-02-20 17:29:32.000000000 -0800
+++ uninstall_pgtap.sql	2009-02-20 17:30:34.000000000 -0800
@@ -50,11 +50,6 @@
 DROP FUNCTION has_user( NAME );
 DROP FUNCTION has_user( NAME, TEXT );
 DROP FUNCTION _is_super( NAME );
-DROP FUNCTION hasnt_role( NAME );
-DROP FUNCTION hasnt_role( NAME, TEXT );
-DROP FUNCTION has_role( NAME );
-DROP FUNCTION has_role( NAME, TEXT );
-DROP FUNCTION _has_role( NAME );
 DROP FUNCTION hasnt_enum( NAME );
 DROP FUNCTION hasnt_enum( NAME, TEXT );
 DROP FUNCTION hasnt_enum( NAME, NAME );
