--- uninstall_pgtap.sql.saf	2009-03-28 15:58:36.000000000 -0400
+++ uninstall_pgtap.sql	2009-03-28 15:58:43.000000000 -0400
@@ -90,11 +94,6 @@
 DROP FUNCTION has_user( NAME );
 DROP FUNCTION has_user( NAME, TEXT );
 DROP FUNCTION _is_super( NAME );
-DROP FUNCTION hasnt_role( NAME );
-DROP FUNCTION hasnt_role( NAME, TEXT );
-DROP FUNCTION has_role( NAME );
-DROP FUNCTION has_role( NAME, TEXT );
-DROP FUNCTION _has_role( NAME );
 DROP FUNCTION hasnt_enum( NAME );
 DROP FUNCTION hasnt_enum( NAME, TEXT );
 DROP FUNCTION hasnt_enum( NAME, NAME );
