--- sql/uninstall_pgtap.sql.saf	2011-11-10 16:30:54.000000000 -0800
+++ sql/uninstall_pgtap.sql	2011-11-10 16:35:13.000000000 -0800
@@ -365,11 +365,6 @@
 DROP FUNCTION has_user( NAME );
 DROP FUNCTION has_user( NAME, TEXT );
 DROP FUNCTION _has_user( NAME );
-DROP FUNCTION hasnt_role( NAME );
-DROP FUNCTION hasnt_role( NAME, TEXT );
-DROP FUNCTION has_role( NAME );
-DROP FUNCTION has_role( NAME, TEXT );
-DROP FUNCTION _has_role( NAME );
 DROP FUNCTION hasnt_enum( NAME );
 DROP FUNCTION hasnt_enum( NAME, TEXT );
 DROP FUNCTION hasnt_enum( NAME, NAME );
@@ -713,3 +708,7 @@
 DROP FUNCTION textarray_text(text[]);
 DROP CAST (boolean AS char(1));
 DROP FUNCTION booltext(boolean);
+DROP CAST (int2vector AS int[]);
+DROP FUNCTION int2vint(int2vector);
+DROP CAST (oidvector AS regtype[]);
+DROP FUNCTION oidvregtype(oidvector);
