--- sql/uninstall_pgtap.sql
+++ sql/uninstall_pgtap.sql
@@ -490,10 +490,6 @@
 DROP FUNCTION has_role( NAME );
 DROP FUNCTION has_role( NAME, TEXT );
 DROP FUNCTION _has_role( NAME );
-DROP FUNCTION enum_has_labels( NAME, NAME[] );
-DROP FUNCTION enum_has_labels( NAME, NAME[], TEXT );
-DROP FUNCTION enum_has_labels( NAME, NAME, NAME[] );
-DROP FUNCTION enum_has_labels( NAME, NAME, NAME[], TEXT );
 DROP FUNCTION hasnt_enum( NAME );
 DROP FUNCTION hasnt_enum( NAME, TEXT );
 DROP FUNCTION hasnt_enum( NAME, NAME );
@@ -824,9 +820,6 @@
 DROP FUNCTION is (anyelement, anyelement, text);
 DROP FUNCTION ok ( boolean );
 DROP FUNCTION ok ( boolean, text );
-DROP FUNCTION diag( VARIADIC anyarray );
-DROP FUNCTION diag( VARIADIC text[] );
-DROP FUNCTION diag ( msg anyelement );
 DROP FUNCTION diag ( msg text );
 DROP FUNCTION finish ();
 DROP FUNCTION _finish ( INTEGER, INTEGER, INTEGER);
@@ -849,3 +842,15 @@
 DROP FUNCTION os_name();
 DROP FUNCTION pg_version_num();
 DROP FUNCTION pg_version();
+DROP CAST (regtype AS text);
+DROP FUNCTION regtypetext(regtype);
+DROP OPERATOR <> ( name[], name[] );
+DROP FUNCTION namearray_ne( name[], name[] );
+DROP OPERATOR = ( name[], name[] );
+DROP FUNCTION namearray_eq( name[], name[] );
+DROP CAST (name[] AS text);
+DROP FUNCTION namearray_text(name[]);
+DROP CAST (text[] AS text);
+DROP FUNCTION textarray_text(text[]);
+DROP CAST (boolean AS char(1));
+DROP FUNCTION booltext(boolean);
