diff --git a/include/boost/random/detail/integer_log2.hpp b/include/boost/random/detail/integer_log2.hpp
index 248243a4b..2e49f281f 100644
--- a/boost/random/detail/integer_log2.hpp
+++ b/boost/random/detail/integer_log2.hpp
@@ -16,7 +16,7 @@

 #include <boost/config.hpp>
 #include <boost/limits.hpp>
-#include <boost/pending/integer_log2.hpp>
+#include <boost/integer/integer_log2.hpp>

 namespace boost {
 namespace random {
