--- examples/interface.cc 2014-05-23 08:01:26.000000000 +0400
+++ examples/interface.cc 2018-01-25 09:24:50.000000000 +0300
@@ -28,14 +28,6 @@

 using namespace crcutil;

-#if defined(__GNUC__)
-// Suppress 'invalid access to non-static data member ...  of NULL object'
-#undef offsetof
-#define offsetof(TYPE, MEMBER) (reinterpret_cast <size_t> \
-    ((&reinterpret_cast <const char &>( \
-        reinterpret_cast <const TYPE *>(1)->MEMBER))) - 1)
-#endif  // defined(__GNUC__)
-
 namespace crcutil_interface {

 template<typename CrcImplementation, typename RollingCrcImplementation>
