--- a/src/debugallocation.cc 2014-04-13 05:35:40.000000000 +0400
+++ b/src/debugallocation.cc 2018-01-28 09:56:21.000000000 +0300
@@ -529,7 +529,7 @@
       b = (MallocBlock*) do_malloc(real_malloced_size(size));
     }
 #else
-    b = (MallocBlock*) do_malloc(real_malloced_size(size);
+    b = (MallocBlock*) do_malloc(real_malloced_size(size));
 #endif

     // It would be nice to output a diagnostic on allocation failure
