diff --git a/lib/Driver/ToolChains.cpp b/lib/Driver/ToolChains.cpp
index 4688335..9173714 100644
--- a/tools/clang/lib/Driver/ToolChains.cpp
+++ b/tools/clang/lib/Driver/ToolChains.cpp
@@ -1060,8 +1060,10 @@ Generic_GCC::GCCInstallationDetector::GCCInstallationDetector(
     Prefixes.push_back(D.InstalledDir + "/..");
 
     // And finally in /usr.
-    if (D.SysRoot.empty())
+    if (D.SysRoot.empty()) {
+      Prefixes.push_back("/opt/rh/devtoolset-3/root/usr");
       Prefixes.push_back("/usr");
+    }
   }
 
   // Loop over the various components which exist and select the best GCC
