--- a/snappy-stubs-public.h.in 2015-07-07 08:45:04.000000000 +0000
+++ b/snappy-stubs-public.h.in 2018-01-22 18:36:52.559809976 +0000
@@ -81,8 +81,8 @@
 typedef std::string string;

 #define DISALLOW_COPY_AND_ASSIGN(TypeName) \
-  TypeName(const TypeName&);               \
-  void operator=(const TypeName&)
+  TypeName(const TypeName&) = delete;      \
+  void operator=(const TypeName&) = delete

 #if !@ac_cv_have_sys_uio_h@
 // Windows does not have an iovec type, yet the concept is universally useful.
